### DISCUSSION TITLE
Errors trying to change timelines

### DISCUSSION THREAD

**Discussion by a-senra** [dialogic-godot/dialogic#2088] [Q&A] [answered]
Hello! First of all, I want to clarify that I'm very new to coding (this week I managed to finally understand autoload and scenes instances) so I apologize if I don't understand correctly the solution you are proposing.

Now, my problem: I have a .gd that manages my timelines in autoload. I have two timelines in array:
```
var current_timeline = 0
var timelines = [
	'00_timeline',
	'01_timeline'
]
```
I'm doing this in order to be able to manage various timelines in the future. Now, this is the rest of the code:

```
func _ready():
	# Connect dialogic signals to functions
	Dialogic.timeline_ended.connect(_ended)
	Dialogic.signal_event.connect(_nextTimeline)

func _ended():
	print("ended")

func _nextTimeline(argument:String):
	# sums 1 if the argument is valid
	if argument =="next_please":
		print ("connected")
		current_timeline += 1 
		# opens the next timeline
		_startTimeline()

func _startTimeline():
	# timeline opener
	Dialogic.start(timelines[current_timeline])
```
The end signal is there because I want to return to the main menu, but it's not implemented.

When this function operates, my run suddenly closes and gives me an error that says: "Out of bounds get index '-1' (on base: 'Array'), and opens the subsystem_voice.gd:37.
```
func is_voiced(index:int) -> bool:
	if dialogic.current_timeline_events[index] is DialogicTextEvent:
		if dialogic.current_timeline_events[index-1] is DialogicVoiceEvent:
			return true
	return false
```

The crash also happens in the middle of the timeline, and the signal is connected, is the Dialogic.start() what crashes my game. 

I guess I'm not using the proper method, so if you can tell me your way of opening a timeline at the end of one I would be very grateful!

**Comment by Jowan-Spooner** [comment]
This is probably a bug. What version are you using?

**Reply by a-senra** [reply]
I'm using Godot 4.2.1, and Dialogic 2.0-Alpha-13

**Comment by CakeVR** [comment]
Hello, is your timeline playing any voices? Could you share it here?

**Reply by a-senra** [reply]
hello, currently no. There is only text and a choice to bring you to the next timeline (where the crash happens). I have not fiddled with styles, glossaries or dialogic variables,  and the character only has a name and the default typing sound. 
Github isn't letting me upload .dtl files, but I have here the project (some text may be in spanish, but it's not important)
[WAGOS.zip](https://github.com/dialogic-godot/dialogic/files/14271436/WAGOS.zip)

**Reply by CakeVR** [reply]
Can you try this PR? 
https://github.com/dialogic-godot/dialogic/pull/2089

If you are unsure how to do that, you may also just copy the code change, it's a few lines.

**Reply by a-senra** [reply]
I copied the code (sorry, this is algo my first time in github, so I was a bit intimidated!) and is the same error, but now in subsystem_choices.gd:189
```
func is_question(index:int) -> bool:
	if dialogic.current_timeline_events[index] is DialogicTextEvent:
		if len(dialogic.current_timeline_events)-1 != index:
			if dialogic.current_timeline_events[index+1] is DialogicChoiceEvent:
				return true

	if dialogic.current_timeline_events[index] is DialogicChoiceEvent:
		if index != 0 and dialogic.current_timeline_events[index-1] is DialogicEndBranchEvent:
			if dialogic.current_timeline_events[dialogic.current_timeline_events[index-1].find_opening_index()] is DialogicChoiceEvent:
				return false
			else:
				return true
		else:
			return true
	return false
```
I'm reattaching the project in case I didn't do it right, thank you so much for the help!
[WAGOS_02.zip](https://github.com/dialogic-godot/dialogic/files/14271921/WAGOS_02.zip)

**Reply by CakeVR** [reply]
Gotcha, then the problem must be a fundamental one and my PR may not even be needed.
Thanks a lot for bringing this to our attention!

**Reply by a-senra** [reply]
Thank you for the quick response! I'll wait patiently for the bug to be fixed, fantastic work :)

**Reply by CakeVR** [reply, marked as answer]
Technically, you could use the `jump` event for now and jump to the next timeline. That will bypass this problem for now.

**Reply by a-senra** [reply]
Oh, thats true. Thanks!

**Reply by CakeVR** [reply]
Another approach you can do is call `Dialogic.end_timeline()` manually before you start the next timeline.
I do think opening multiple timeline should work in theory, but this crashing may be a bug.

**Reply by a-senra** [reply]
okay, I just tried that and I got a different error. What I did was call a function for ending the current timeline when the signal "next_please" is activated:

```
func _nextTimeline(argument:String):
	# Si el string coincide, se pasa a la siguiente del array
	if argument =="next_please":
		print ("connected")
		_deleteTimeline()
		current_timeline =+ 1 
		# Se abre la siguiente timeline
		_startTimeline()
```
An error pops up in the DialogicGameHandler.gd:278, in the function clear: "Invalid call. Nonexistent function 'clean' in base 'Nil'".

```
func clear(clear_flags:=ClearFlags.FULL_CLEAR) -> bool:

	if !clear_flags & ClearFlags.TIMLEINE_INFO_ONLY:
		for subsystem in get_children():
			if subsystem is DialogicSubsystem:
				(subsystem as DialogicSubsystem).clear_game_state(clear_flags)

	# Resetting variables
	if current_timeline:
		# This is necessary because otherwise INTERNAL GODOT ONESHOT CONNECTIONS
		# are disconnected before they can disconnect themselves.
		await get_tree().process_frame
		current_timeline.clean()
```
Right now I'm doing good using jumps, but if you need me to try more things in order to solve the bug I'm happy to help :)

**Comment by CakeVR** [comment]
@Jowan-Spooner has solved this bug, please try the latest commit on the `main` branch. Thanks for your patience!
